### PR TITLE
Allow a number of broadcast transport actions to resolve data streams

### PR DESCRIPTION
**Pull request by martijnvg**
Change `TransportBroadcastByNodeAction` and `TransportBroadcastReplicationAction`
to be able to resolve data streams by default. Implementations can change this ability.

This change allows to following APIs to resolve data streams: flush,
refresh (already supported data streams), force merge, clear indices cache,
indices stats, segments, upgrade stats, upgrade, validate query,
searchable snapshots stats, clear searchable snapshots cache and
reload analyzers APIs.

Relates to #53100

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 1]
```diff
@@ -446,6 +449,14 @@ public static IndicesOptions strictExpand() {
         return STRICT_EXPAND_OPEN_CLOSED;
     }
 
+    /**
+     * @return indices option that requires every specified index to exist, expands wildcards to both open and closed
+     * indices and allows that no indices are resolved from wildcard expressions (not returning an error).
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java:454]: Maybe add that data streams are included to distinguish it from `::strictExpand()`

**Review by danhermann — APPROVED** [review]
LGTM. Left one minor comment

**Comment by henningandersen** [comment]
I think `ForgetFollowerAction.Request` also cannot meaningfully resolve data streams.

I am also a bit worried that we here are introducing support for data streams in a wide array of APIs without any tests for it. I think it would be better to tackle them one by one, changing the indices options in each specific request and adding the necessary tests for it?

**Comment by martijnvg** [comment]
> I think ForgetFollowerAction.Request also cannot meaningfully resolve data streams.

True and it doesn't support aliases and wildcard expressions too. The corresponding transport action uses the leader and follower indices as provided in the api. Which why this PR doesn't add data stream support to this api either. The `ForgetFollowerAction.Request` supplies the leader index to the base class (`BroadcastRequest`) and that is then used to find the leader shards, but in the transport action's `shardOperation()` method it then uses the leader index as is provided via api. Using using anything other than the concrete leader index results in failure. I think this behaviour is intentional.

> I think it would be better to tackle them one by one, changing the indices options in each specific request and adding the necessary tests for it?

I've added tests for all the APIs that have a request that extends from `BroadcastRequest` and that support data streams. So that should be ok? Or do you prefer that I revert the change to `BroadcastRequest` and make the change to all the concrete request classes that should support data streams? (similar to the refresh request class prior to this PR)

**Comment by henningandersen** [comment]
About `ForgetFollowerAction`, I think it will result in an NPE

```
        final Index leaderIndex = clusterService.state().metadata().index(request.leaderIndex()).getIndex();
```

I think we should rather use an indicesOptions that does not support data streams. I think the investigation work you are doing will likely address this?

I think the tests look fine (though the forget follower action is missing). Thanks for pointing that out.

I wonder if I should review this now or wait until you concluded the experiment on extracting the data stream option?

**Comment by martijnvg** [comment]
> I wonder if I should review this now or wait until you concluded the experiment on extracting the data stream option?

Yes, I think we should wait with this pr and see how #56034 unfolds.

**Comment by martijnvg** [comment]
I've updated this  PR based on the changes that came with #56034.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**